src/main/java/com/google/idea/perf/util/LruCache.kt

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.idea.perf.util
+
+private class Node<K:Any, V: Any>(var key: K, var value: V) {
+    var next: Node<K, V>? = null
+    var previous: Node<K, V>? = null
+}
+
+class LruCache<K: Any, V: Any>(private val maxCapacity: Int = 0) {
```

Comment:
This is neat, but there's no need to implement it ourselves!
There is already `com.intellij.util.containers.SLRUMap` (and the `SLRUCache` subclass).

Alternatively, you can even implement an LRU cache via `LinkedHashMap`. Search for "LRU" in https://docs.oracle.com/javase/8/docs/api/java/util/LinkedHashMap.html, and see example usage in `com.intellij.openapi.vfs.newvfs.impl.UserDataInterner#ourCache`:
```java
ourCache = new LinkedHashMap<...>(20, true) {
  @Override
  protected boolean removeEldestEntry(Map.Entry<...> eldest) {
    return size() > 15;
  }
};
```